src/optimizer.cpp

```diff
@@ -124,67 +103,104 @@ geometry_msgs::msg::TwistStamped Optimizer::evalControl(
   const geometry_msgs::msg::Twist & robot_speed,
   const nav_msgs::msg::Path & plan, nav2_core::GoalChecker * goal_checker)
 {
-  for (size_t i = 0; i < settings_.iteration_count_; ++i) {
-    generated_trajectories_ =
-      generateNoisedTrajectories(robot_pose, robot_speed);
-    auto && costs = critic_manager_.evalTrajectoriesScores(
-      state_, generated_trajectories_, plan, robot_pose, goal_checker);
-    updateControlSequence(costs);
-  }
+  prepare(robot_pose, robot_speed, plan, goal_checker);
 
-  auto control = getControlFromSequenceAsTwist(
-    settings_.control_sequence_shift_offset_,
-    plan.header.stamp);
+  do {
+    optimize();
+  } while (fallback(critics_data_.fail_flag));
+
+  auto control = getControlFromSequenceAsTwist(plan.header.stamp);
+
+  if (settings_.shift_control_sequence) {
+    shiftControlSequence();
+  }
 
-  shiftControlSequence();
   return control;
 }
 
-void Optimizer::shiftControlSequence()
+void Optimizer::optimize()
 {
-  if (settings_.control_sequence_shift_offset_ == 0) {
-    return;
+  for (size_t i = 0; i < settings_.iteration_count; ++i) {
+    generated_trajectories_ = generateNoisedTrajectories();
```

Comment:
Due to the subtly of the references in the `critics_data_` on the class members, I feel like we should make some comment, somewhere, that the critics_data_ holds a reference to the path/state/trajectories so that we don't need to update explicitly since they're stored class members. `Prepare()` explicitly sets the path and state, so its easy for someone not to even notice it, since the trajectories are the only thing changing after `prepare()` is called and is never overrided. 

My personal preference would be to add a member to set the generated trajectories in the loop to `critics_data_`, but a comment would do as well.

Reply:
I'm not sure that we need to add something. I agree that struct that keeps references is probably rare stuff. 

If anyone would be interested in what data passed to critic manager he would come to CriticFunctionData, so he **must** notice references there.

Reply:
I agree its not a _need_, but its nice to leave people breadcrumbs so they don't spin their wheels if we know we're doing something unusual or particularly subtle

Reply:
added comment on the member var